src/tools/arp-scan.c

```diff
@@ -0,0 +1,484 @@
+		pthread_exit(NULL);
+	}
+	if(arp_verbose)
+		printf("Scanning interface %s (%s/%i)...\n", iface, ipstr, cidr);
+
+	// Create socket for ARP communications
+	const int arp_socket = create_arp_socket(ifindex, iface);
+
+	// Cannot create socket, likely a permission error
+	if(arp_socket < 0)
+		pthread_exit(NULL);
+
+	// Get hardware address of client machine
+	unsigned char mac[16] = { 0 };
+	get_hardware_address(arp_socket, iface, mac);
+
+	// Define destination IP address by masking source IP with netmask
+	struct in_addr dst_addr = { 0 };
+	dst_addr.s_addr = src_addr.sin_addr.s_addr & mask.sin_addr.s_addr;
+
+	// Allocate memory for ARP response buffer
+	const size_t arp_result_len = 1 << (32 - cidr);
+	struct arp_result *result = calloc(arp_result_len, sizeof(struct arp_result));
+
+	for(unsigned int scan_id = 0; scan_id < NUM_SCANS; scan_id++)
+	{
+#ifdef DEBUG
+		printf("Scanning interface %s (%s/%i) for the %i. time\n", iface, ipstr, cidr, scan_id + 1);
+#endif
+		// Send ARP requests to all IPs in subnet
+		if(send_arps(arp_socket, ifindex, iface, mac, &src_addr.sin_addr, dst_addr, cidr) != 0)
+			break;
+
+		// Read ARP responses
+		if(read_arp(arp_socket, iface, &dst_addr, result, arp_result_len, scan_id) != 0)
+			break;
+	}
+
+	// Check if there are any results
+	unsigned int replies = 0;
+	for(unsigned int i = 0; i < arp_result_len; i++)
+		for(unsigned int j = 0; j < NUM_SCANS; j++)
+			replies += result[i].replied[j];
+
+	if(pthread_mutex_lock(&lock) != 0)
+		return NULL;
+
+	if(replies == 0)
+	{
+		printf("No devices found on interface %s (%s/%i)\n", iface, ipstr, cidr);
+		goto arp_scan_iface_end;
+	}
+
+	// Print results
+	printf("ARP scan on interface %s (%s/%i) finished\n", iface, ipstr, cidr);
+	printf("%-20s %-16s %-17s  Reply matrix\n", "IP address", "Interface", "MAC address");
+	for(unsigned int i = 0; i < arp_result_len; i++)
+	{
+		// Check if IP address replied
+		bool replied = false, multiple_replies = false;
+		for(unsigned int j = 0; j < NUM_SCANS; j++)
+		{
+			if(result[i].replied[j] > 0)
+			{
+				replied = true;
+				multiple_replies |= result[i].replied[j] > 1;
+			}
+		}
+		if(!replied)
+			continue;
+
+		// Convert IP address to string
+		struct in_addr ip = { 0 };
+		ip.s_addr = htonl(ntohl(dst_addr.s_addr) + i);
+		inet_ntop(AF_INET, &ip, ipstr, INET_ADDRSTRLEN);
+
+		// Print MAC addresses
+		unsigned int j = 0;
+		for(j = 0; j < MAX_MACS; j++)
+		{
+			// Check if result[i].mac[j] is all-zero
+			if(memcmp(result[i].mac[j], "\x00\x00\x00\x00\x00\x00", 6) == 0)
+				break;
+
+			// Print MAC address
+			printf("%-20s %-16s %02x:%02x:%02x:%02x:%02x:%02x ",
+			       ipstr, iface,
+			       result[i].mac[j][0],
+			       result[i].mac[j][1],
+			       result[i].mac[j][2],
+			       result[i].mac[j][3],
+			       result[i].mac[j][4],
+			       result[i].mac[j][5]);
+
+			for(unsigned int k = 0; k < NUM_SCANS; k++)
+			{
+				printf(" %s", result[i].replied[k] > 0 ? "X" : "-");
```

Comment:
I see what you mean with the X being rather a mark for a failure. However, `+` or `*` are good alternatives because they both look rather similar to `-` and one would really have to pay attention to actually see a difference whereas it was much easier before.

Maybe `✓` ?

Reply:
There are many options, but personal taste is... well, personal and there's nothing wrong about that.
My comment was intended only to avoid (a possible) confusion caused by `X`.

I only have one preference about this: I didn't like the idea to use "nothing" (space) for failure. This could cause confusion if the test fails for all attempts (nothing will show up).

From a "technical" point of view, we need 2 visually distinct characters to represent Success vs Fail.

We can test other characters like:
```
192.168.2.3          enp2s0           22:a6:2f:14:4a:89  ■ ■ ■ ■ ■ ■ ■ ■ ■ ■
192.168.2.4          enp2s0           32:3a:44:34:4f:1f  ■ · · · ■ ■ ■ ■ ■ ·
192.168.2.107        enp2s0           c8:c9:a3:23:18:f3  · · · · · ■ · · ■ ·

ASCII chars 250 and 254
```

Or we can use any characters and add some kind of "legend":
```
ARP result (■=success; ·=failure):
192.168.2.3          enp2s0           22:a6:2f:14:4a:89  ■ ■ ■ ■ ■ ■ ■ ■ ■ ■
192.168.2.4          enp2s0           32:3a:44:34:4f:1f  ■ · · · ■ ■ ■ ■ ■ ·
192.168.2.107        enp2s0           c8:c9:a3:23:18:f3  · · · · · ■ · · ■ ·
```

Reply:
Marking this resolved, since the _Reply matrix_ was replaced by _Reply Rate_ column.